### PR TITLE
Expose interfaces to assist with creating CEL evaluation decorators

### LINKED ISSUE #363: Support external CEL program decorators
**Feature request checklist**

- [X] There are no issues that match the desired change
- [X] The change is large enough it can't be addressed with a simple Pull Request
- [ ] If this is a bug, please file a [Bug Report](./ISSUE_TEMPLATE).

**Change**
The `cel.Program` creation step can accept a set of `InterpretableDecorator` instances which
are capable of observing, altering, or replacing `Interpretable` steps in the program plan. This
is useful for evaluating expressions whose values are all constants, or for tracking the set of
functions and fully qualified attributes which may appear program plan. There are a number
of application specific use cases which developers would like to support, but which should not be part of the general purpose CEL library.

### PR DESCRIPTION
**Pull request by TristonianJones**
If the `cel.EvalOptions(cel.OptOptimize)` option is provided to the `env.Program(ast, opts)` call, CEL applies a series of `InterpretableDecorator` functional objects as part of the program plan stage which in some cases will replace or rewrite CEL expressions to perform more optimally. 

CEL also uses the same technique to support state tracking and exhaustive evaluation. This is a generally useful technique, but unless the user had access to the internal implementation details of CEL, it was impossible to use outside of CEL. The following interface changes make it possible for a limited set of custom decorators to be created where any decorator may inspect any `Attribute` or `InterpretableConst` or `InterpretableCall` at program plan phase and observe, alter, or replace the evaluation node to support a specific use case.

Closes #363

### CODE REVIEW

**Comment by TristonianJones** [comment]
@guptadipanshu, @mtheodorides FYI. I think this is roughly what I'd talked to each of you about in terms of making it possible to interrogate or alter CEL evaluation to meet your needs.

**Comment by guptadipanshu** [comment]
thanks @TristonianJones for creating it :)

File: interpreter/interpretable.go [hunk 1]
```diff
@@ -32,6 +34,48 @@ type Interpretable interface {
...
+
+	// Attr returns the Attribute value.
+	Attr() Attribute
+
+	// Adapter returns the type adapter to be used for adapting resolved Attribute values.
+	Adapter() ref.TypeAdapter
+
+	// AddQualifier proxies the Attribute.AddQualifier method.
+	// Note, this method will mutate the current attribute state. If the desire is to clone the
+	// Attribute, then alte
```
Comment (JimLarson) [inline interpreter/interpretable.go]: The sentence trails o

**Review by JimLarson — APPROVED** [review]

File: interpreter/interpretable.go (hunk 1 above)
Comment (TristonianJones) [reply interpreter/interpretable.go]: 🥇 For such a well-phrased commen

**Pull request merged**